azurerm/resource_arm_role_definition.go

```diff
@@ -163,7 +164,8 @@ func resourceArmRoleDefinitionDelete(d *schema.ResourceData, meta interface{}) e
 	client := meta.(*ArmClient).roleDefinitionsClient
 	ctx := meta.(*ArmClient).StopContext
 
-	roleDefinitionId := d.Get("role_definition_id").(string)
+	parts := strings.Split(d.Id(), "/")
```

Comment:
@katbyte @tombuildsstuff , what is the point of introducing the `role_definition_id` if we already has the globally `id` attribute?

I would like to suggest deprecating it using the [similar syntax here](https://github.com/terraform-providers/terraform-provider-azurerm/blob/670b1b93238dbf646dab956955fc708159327659/azurerm/resource_arm_dns_cname_record.go#L41). But if we have to keep it, I would also suggest the following logic:

```go
roleDefId := d.Get("role_definition_id").(string)
if roleDefId == "" {
  id, err := parseAzureResourceID(d.Id())
  if err != nil { ... }
  roleDefId = id.Path[...]
}
```

Reply:
@JunyiYi in this case we're going to need to update the ID to contain all of the required information - [see here](https://github.com/terraform-providers/terraform-provider-azurerm/pull/1816/commits/12b06b6f0140cb8e385131af47d85989e698d3aa) for an example of how we're doing this with the storage resources. In general we shouldn't be accessing the config in the Read and Delete functions since it can be present (at Import time during the Read function, and then the associated Terraform Config's been deleted in the Delete function)